### ISSUE TITLE
About the results on VOC with 92 examples

### ISSUE THREAD

**Issue by Jin-Ying** [Haochen-Wang409/U2PL#19]
Thank you for your expressive work and delicate code. When reproducing the results on VOC, I can reproduce the results with 183, 366, 732, 1464 examples, but only get about 52 with 92 examples, while the reported result is 67.98. Can you offer some help? Sincere gratitude to you.

**Comment by Haochen-Wang409** [comment]
Could you provice the ```config.yaml``` you used?

**Comment by Jin-Ying** [comment]
dataset: # Required.
  type: pascal_semi
  train:
    data_root: ../../../../data/VOC2012
    data_list: ../../../../data/splits/pascal/92/labeled.txt
    flip: True
    GaussianBlur: False
    rand_resize: [0.5, 2.0]
    #rand_rotation: [-10.0, 10.0]
    crop:
      type: rand
      size: [513, 513] # crop image with HxW size
  val:
    data_root: ../../../../data/VOC2012
    data_list: ../../../../data/splits/pascal/val.txt
    crop:
      type: center
      size: [513, 513] # crop image with HxW size
  batch_size: 4
  n_sup: 92
  noise_std: 0.1
  workers: 2
  mean: [123.675, 116.28, 103.53]
  std: [58.395, 57.12, 57.375]
  ignore_label: 255

trainer: # Required.
  epochs: 80
  eval_on: True
  optimizer:
    type: SGD
    kwargs:
      lr: 0.001  # 4GPUs
      momentum: 0.9
      weight_decay: 0.0001
  lr_scheduler:
    mode: poly
    kwargs:
      power: 0.9
  unsupervised:
    TTA: False
    drop_percent: 80
    apply_aug: cutmix
  contrastive:
    negative_high_entropy: True
    low_rank: 3
    high_rank: 20
    current_class_threshold: 0.3
    current_class_negative_threshold: 1
    unsupervised_entropy_ignore: 80
    low_entropy_threshold: 20
    num_negatives: 50
    num_queries: 256
    temperature: 0.5

saver:
  snapshot_dir: checkpoints
  pretrain: ''

criterion:
  type: CELoss
  kwargs:
    use_weight: False

net: # Required.
  num_classes: 21
  sync_bn: True
  ema_decay: 0.99
  encoder:
    type: u2pl.models.resnet.resnet101
    kwargs:
      multi_grid: True
      zero_init_residual: True
      fpn: True
      replace_stride_with_dilation: [False, True, True]  #layer0...1 is fixed, layer2...4
  decoder:
    type: u2pl.models.decoder.dec_deeplabv3_plus
    kwargs:
      inner_planes: 256
      dilations: [12, 24, 36]

**Comment by Haochen-Wang409** [comment]
Did you train the model with 4 GPUs?

**Comment by Jin-Ying** [comment]
Yes, with 4 GPUs.

**Comment by Haochen-Wang409** [comment]
The following confg might help.
```
dataset: # Required.
  type: pascal_semi
  train:
    data_root: ../../../../data/VOC2012
    data_list: ../../../../data/splits/pascal/1464/labeled.txt
    flip: True
    GaussianBlur: False
    rand_resize: [0.5, 2.0]
    #rand_rotation: [-10.0, 10.0]
    crop:
      type: rand
      size: [513, 513] # crop image with HxW size
  val:
    data_root: ../../../../data/VOC2012
    data_list: ../../../../data/splits/pascal/val.txt
    crop:
      type: center
      size: [513, 513] # crop image with HxW size
  batch_size: 4
  n_sup: 1464
  noise_std: 0.1
  workers: 2
  mean: [123.675, 116.28, 103.53]
  std: [58.395, 57.12, 57.375]
  ignore_label: 255

trainer: # Required.
  epochs: 80
  eval_on: True
  optimizer:
    type: SGD
    kwargs:
      lr: 0.001  # 4GPUs
      momentum: 0.9
      weight_decay: 0.0001
  lr_scheduler:
    mode: poly
    kwargs:
      power: 0.9
  unsupervised:
    TTA: False
    drop_percent: 80
    apply_aug: cutmix
  contrastive:
    loss_weight: 0.1
    negative_high_entropy: True
    low_rank: 3
    high_rank: 20
    current_class_threshold: 0.3
    current_class_negative_threshold: 1
    unsupervised_entropy_ignore: 80
    low_entropy_threshold: 20
    num_negatives: 50
    num_queries: 256
    temperature: 0.5

saver:
  snapshot_dir: checkpoints
  pretrain: ''

criterion:
  type: CELoss
  kwargs:
    use_weight: False

net: # Required.
  num_classes: 21
  sync_bn: True
  ema_decay: 0.99
  encoder:
    type: u2pl.models.resnet.resnet101
    kwargs:
      multi_grid: True
      zero_init_residual: True
      fpn: True
      replace_stride_with_dilation: [False, True, True]  #layer0...1 is fixed, layer2...4
  decoder:
    type: u2pl.models.decoder.dec_deeplabv3_plus
    kwargs:
      inner_planes: 256
      dilations: [12, 24, 36]
```

**Comment by Haochen-Wang409** [comment]
U<sup>2</sup>PL with ```1464``` labeled images is robust against different ```loss_weight``` in contrastive learning, while ```92``` labeled images may cause the instability during training.

**Comment by Jin-Ying** [comment]
Now I get 59.9 on 92 examples, which is still far below the reported results in your original paper. Here is my current config (which only replaces 1464 with 92 in your config). Sincere gratitude.

dataset: # Required.
  type: pascal_semi
  train:
    data_root: ../../../../data/VOC2012
    data_list: ../../../../data/splits/pascal/92/labeled.txt
    flip: True
    GaussianBlur: False
    rand_resize: [0.5, 2.0]
    #rand_rotation: [-10.0, 10.0]
    crop:
      type: rand
      size: [513, 513] # crop image with HxW size
  val:
    data_root: ../../../../data/VOC2012
    data_list: ../../../../data/splits/pascal/val.txt
    crop:
      type: center
      size: [513, 513] # crop image with HxW size
  batch_size: 4
  n_sup: 92
  noise_std: 0.1
  workers: 2
  mean: [123.675, 116.28, 103.53]
  std: [58.395, 57.12, 57.375]
  ignore_label: 255

trainer: # Required.
  epochs: 80
  eval_on: True
  optimizer:
    type: SGD
    kwargs:
      lr: 0.001  # 4GPUs
      momentum: 0.9
      weight_decay: 0.0001
  lr_scheduler:
    mode: poly
    kwargs:
      power: 0.9
  unsupervised:
    TTA: False
    drop_percent: 80
    apply_aug: cutmix
  contrastive:
    loss_weight: 0.1
    negative_high_entropy: True
    low_rank: 3
    high_rank: 20
    current_class_threshold: 0.3
    current_class_negative_threshold: 1
    unsupervised_entropy_ignore: 80
    low_entropy_threshold: 20
    num_negatives: 50
    num_queries: 256
    temperature: 0.5

saver:
  snapshot_dir: checkpoints
  pretrain: ''

criterion:
  type: CELoss
  kwargs:
    use_weight: False

net: # Required.
  num_classes: 21
  sync_bn: True
  ema_decay: 0.99
  encoder:
    type: u2pl.models.resnet.resnet101
    kwargs:
      multi_grid: True
      zero_init_residual: True
      fpn: True
      replace_stride_with_dilation: [False, True, True]  #layer0...1 is fixed, layer2...4
  decoder:
    type: u2pl.models.decoder.dec_deeplabv3_plus
    kwargs:
      inner_planes: 256
      dilations: [12, 24, 36]

**Comment by Haochen-Wang409** [comment]
However, I trained with this ```config.yaml``` and have got **69.29** on epoch 20...

I am not sure what's the problem since I cannot reproduce your performance... :(

**Comment by Haochen-Wang409** [comment]
What kind of GPU did you use for training?

**Comment by Jin-Ying** [comment]
In your config.yaml, the label #num is 1464, did you simply replace it with 92?

**Comment by Haochen-Wang409** [comment]
Yes, of course. I replace ```1464``` in ```data_list``` and ```n_sup``` to ```92```.

**Comment by Jin-Ying** [comment]
I use 4*A100 for training.

**Comment by Haochen-Wang409** [comment]
Could you provide the training log？
